util/expensivequery/expensivequery.go

```diff
@@ -73,12 +85,214 @@ func (eqh *Handle) Run() {
+// If Performance.ServerMemoryQuota is not set, use system memory usage and total memory * 80% to check oom risk.
+func (eqh *Handle) oomKillerAlert() {
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	if eqh.record.useInstanceMemory {
+		runtime.ReadMemStats(instanceStats)
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, eqh.record.err = memory.MemUsed()
+		if eqh.record.err != nil {
+			logutil.BgLogger().Error("Get system usage memory fail.", zap.Error(eqh.record.err))
+			return
+		}
+	}
+
+	// If we use instance memory usage to check oom risk, we also need use NextGC compared with the quota.
+	// Maybe the instance will oom but the HeapAlloc isn't updated after golang GC.
+	// Go's GC don't take the system memory limit into consideration, that means:
+	//
+	// 1. You have 16G physical memory;
+	// 2. You have 9G live objects;
+	// 3. Go will run GC when memory usage reaches 9G * (1 + GOGC / 100) = 18G, OOM.
+	if float64(memoryUsage) > float64(eqh.record.serverMemoryQuota)*config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio ||
+		(eqh.record.useInstanceMemory && instanceStats.NextGC > eqh.record.serverMemoryQuota) {
+		// At least ten seconds between two recordings that memory usage is less than threshold (default 80% system memory).
+		// If the memory is still exceeded, only records once.
+		if time.Since(eqh.record.lastRecordTime) > 10*time.Second {
+			eqh.oomRecord(memoryUsage)
+		}
+		eqh.record.lastRecordTime = time.Now()
+	}
+}
+
+func (eqh *Handle) oomRecord(memUsage uint64) {
+	if eqh.record.useInstanceMemory {
+		logutil.BgLogger().Warn("The TiDB instance now takes a lot of memory, has the risk of OOM",
+			zap.Any("tidbMemUsage", memUsage),
+			zap.Any("serverMemoryQuota", eqh.record.serverMemoryQuota),
+		)
+	} else {
+		logutil.BgLogger().Warn("The os system now takes a lot of memory, has the risk of OOM",
+			zap.Any("osMemUsage", memUsage),
+			zap.Any("osMemoryTotal", eqh.record.serverMemoryQuota),
+		)
+	}
+
+	if eqh.record.err = disk.CheckAndInitTempDir(); eqh.record.err != nil {
+		return
+	}
+	eqh.oomRecordSQL()
+	eqh.oomRecordProfile()
+
+	tryRemove := func(filename []string) {
+		// Keep the last 5 files
+		if len(filename) < 6 {
+			return
+		}
+		err := os.Remove(filename[0])
+		if err != nil {
+			logutil.BgLogger().Error("Remove temp files failed.", zap.Error(err))
+		}
+		filename = filename[1:]
+	}
+	tryRemove(eqh.record.lastLogFileName)
+	for i := range eqh.record.lastProfileFileName {
+		tryRemove(eqh.record.lastProfileFileName[i])
+	}
+}
+
+func (eqh *Handle) oomRecordSQL() {
+	sm := eqh.sm.Load().(util.SessionManager)
+	processInfo := sm.ShowProcessList()
+	pinfo := make([]*util.ProcessInfo, 0, len(processInfo))
+	for _, info := range processInfo {
+		if len(info.Info) != 0 {
+			pinfo = append(pinfo, info)
+		}
+	}
+	now := time.Now()
+
+	fileName := filepath.Join(config.GetGlobalConfig().TempStoragePath, "oom_sql"+time.Now().Format(time.RFC3339))
```

Comment:
1. oom_sql --> running_sql
2. Is it seems that the time of sql files and profile files are different? If so, we need get the time out of recordSQL and recordProfile

Reply:
Addressed. I set `lastRecordTime` to `time.now()` before doRecord, and convert the time to filename.